contracts/staking/PoSValidatorManager.sol

```diff
@@ -130,4 +146,170 @@ abstract contract PoSValidatorManager is IPoSValidatorManager, ValidatorManager
+            nonce: nonce,
+            startTime: block.timestamp
+        });
+    }
+
+    function initializeEndDelegation(bytes32 validationID) external {
+        PoSValidatorManagerStorage storage $ = _getPoSValidatorManagerStorage();
+
+        // Ensure the delegator is active
+        Delegator memory delegator = $._delegatorStakes[validationID][_msgSender()];
+        require(
+            delegator.status == DelegatorStatus.Active, "ValidatorManager: delegator not active"
+        );
+        delegator.status = DelegatorStatus.PendingRemoved;
+        delegator.endedAt = uint64(block.timestamp);
+
+        $._delegatorStakes[validationID][_msgSender()] = delegator;
+
+        Validator memory validator = _getValidator(validationID);
+        require(validator.weight > delegator.weight, "PoSValidatorManager: Invalid weight");
+        validator.weight -= delegator.weight;
+        _setValidator(validationID, validator);
+
+        // Submit the message to the Warp precompile.
+        uint64 nonce = _getAndIncrementNonce(validationID);
+        bytes memory setValidatorWeightPayload = ValidatorMessages
+            .packSetSubnetValidatorWeightMessage(validationID, nonce, validator.weight);
+        $._pendingEndDelegatorMessages[validationID][_msgSender()] = setValidatorWeightPayload;
+        bytes32 messageID = WARP_MESSENGER.sendWarpMessage(setValidatorWeightPayload);
+
+        emit DelegatorRemovalInitialized({
+            validationID: validationID,
+            setWeightMessageID: messageID,
+            delegator: _msgSender(),
+            validatorWeight: validator.weight,
+            nonce: nonce,
+            endTime: block.timestamp
+        });
+    }
+
+    function resendEndDelegation(bytes32 validationID, address delegator) external {
+        _checkPendingEndDelegatorMessage(validationID, delegator);
+        PoSValidatorManagerStorage storage $ = _getPoSValidatorManagerStorage();
+        WARP_MESSENGER.sendWarpMessage($._pendingEndDelegatorMessages[validationID][delegator]);
+    }
+
+    function completeEndDelegation(uint32 messageIndex, address delegator) external {
+        PoSValidatorManagerStorage storage $ = _getPoSValidatorManagerStorage();
+
+        // Unpack the Warp message
+        WarpMessage memory warpMessage = _getPChainWarpMessage(messageIndex);
+        (bytes32 validationID, uint64 nonce,) =
+            ValidatorMessages.unpackSubnetValidatorWeightUpdateMessage(warpMessage.payload);
+        _checkPendingEndDelegatorMessage(validationID, delegator);
+        delete $._pendingEndDelegatorMessages[validationID][delegator];
+
+        // Update the validator weight and delegator status
+        Validator memory validator = _getValidator(validationID);
+        require(validator.messageNonce >= nonce, "PoSValidatorManager: invalid nonce");
+        $._delegatorStakes[validationID][delegator].status = DelegatorStatus.Completed;
+
+        emit DelegationEnded(validationID, delegator, nonce);
```

Comment:
We should be checking uptime, giving out rewards, and returning stake in here I think?

For uptime, I think anyone should be able to provide an uptime proof for a validator. Since we track the seconds of uptime, a proof can be provided basically at any point, and if the highest reported `uptimeSeconds` is at least 80% of the time from the validation period starting and the MIN(validation end, current time), then we can disburse rewards.

Reply:
I'm going to defer uptime/rewards handling since I think there are still some open design questions here.

Reply:
I implemented uptime handling here: https://github.com/ava-labs/teleporter/pull/520

Happy to merge that into this branch and include it in this PR if you prefer.